doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,364 @@
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // needed to determine timeouts
+ start_time: Instant,
+ // the plan will hold the cluster_id, config, mechanism, etc...
+ plan: AlterClusterPlan,
+ // idempotency key will need to be part of the  message and will need to be
+ // validated to ensure we're looking at the correct plan, in case we
+ // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
+ idempotency_key: Uuid,
+}
+```
+
+In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
+session `ClusterIds` to `PendingClusterReconfigurations`.
+
+*Currently due to limitations in sources and sinks, we will need to maintain
+the disruptive alter mechanism. Once multiple replicas are allowed for sources
+and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.
+
+
+#### Handling PendingClusterReconfigure Messages
+
+The responsibilities of this operation will be:
+1. checking for cancellations
+2. checking for the `WAIT` condition
+3. finalizaing the reconfiguration
+4. returning to the user
+
+__Checking the `WAIT` Condition__
+The message handler will perform all checks on every run. If the wait condition
+has not yet been met, we will reschedule the message to be resent. For each
+reschedule we'll spawn a task that waits some duration and sends a
+`ClusterReconfigureCheckReady` message with the connection_id.
+For `WAIT FOR` we can wait for the exact duration that would cause the next check to succeed.
+For `WAIT UNTIL CAUGHT UP` we can just wait for three seconds.
+
+__Finalize the Reconfiguration__
+Finalizing the reconfiguration will occur once the `wait` condition has been
+met. During this phase, the following steps will occur.
+- Drop any non pending replicas that do not have a billed_as value.
+- Finalize the pending replicas:
+  - Rename to remove the `-pending` suffix.
+  - Alter config to set `pending` to true.
+- Update the cluster catalog to the new desired config.
+
+__Handling Cancelation__
+When a cancel event is read we'll check to see if the active connection has a value
+in a `pending_cluster_reconfiguration` BTreeMap, and if so remove it and clean up leftover state.
+
+
+__v2 coniderations__
+Reverting the reconfiguration
+On failure or timeout of the reconfiguration, this task will revert to the state
+prior to the `ALTER CLUSTER` DDL. It will remove reconfigure replicas, and ensure the
+cluster config matches the original version.
+
+
+#### Recover from Environmentd Crash
+Initially, for v1, we will not recover. We will have to remove all `pending`
+replicas.
+
+For v2, we will need to store the AlterClusterPlan durably for all backgrounded reconfigurations.
+We will then need the coordinator to attempt to continue an ongoing recongiuration on restart.
```

Comment:
There's no need to continue anything here! On restart the catalog will remove the non-pending replicas, promote the pending to non-pending, then proceed with compute controller initialization like normal which will cause the new desired state to take affect.

Reply:
Is that the behavior we want? I imagine if environmentd were to get restarted leading a small amount of downtime we wouldn't want to remove all non-pending fully hydrated replicas and promote replicas that have an unknown hydration state? I would expect the behavior we want to strive for here is to continue with the alter and only perform the finalization once the alter strategy condition is met.

Reply:
My understanding is that on envd restart we always get a whole new set of replicas. The new envd does not hook in to the in-progress hydrating replicas, so they give us no benefit. Recommend verifying this behavior with the cluster team.

Reply:
If envd is restarted but some clusterds are not restarted, we aim to not disturb these clusterds, and reuse them as much as possible. That is, the coordinator [tries](https://github.com/MaterializeInc/materialize/blob/9a1b5a8d24b6da9237fbf546f2f93628ed9d9185/src/adapter/src/coord.rs#L2602) to bootstrap timestamps in a way that existing dataflows on the running clusterds can be reused, and the clusterds [try](https://github.com/MaterializeInc/materialize/blob/5a35bda0ece9b45ea230933b21255e780fbfd2b1/src/compute/src/server.rs#L548) to recognize when compute commands that they receive from a new controller reflect already running things. See also `platform/reconciliation.md`.

Reply:
If the v2 mechanism is not based on rehydration but a parity of the frontiers between old and new replicas then we should be able to handle both the case that existing flows are kept alive and the case where full rehydration is required. In the former we'd wait for new replicas to catch up ,and would only finalize the alter once that had occurred. In the latter, since both sets of replicas had their dataflows restarted, we'd expect to have parity immediately.